src/domain_bridge/wait_for_qos_handler.hpp

```diff
@@ -120,38 +123,51 @@ class WaitForQosHandler
     auto invoke_callback_when_qos_ready = [this, node, cv]()
       {
         auto event = node->get_graph_event();
-        while (!this->shutting_down_.load()) {
+        while (true) {
           // Wait for graph event
           // Note, in case new publishers don't trigger a graph event we add a
           // timeout so that we can still poll periodically for new publishers
           node->wait_for_graph_change(event, std::chrono::seconds(1));
           event->check_and_clear();
-          // Check if QoS is ready for any of the topics
+
           {
             std::unique_lock<std::mutex> lock(waiting_map_mutex_);
+            // If we're shuttdown down, exit the thread
+            if (this->shutting_down_) {
+              return;
+            }
+
+            // Check if QoS is ready for any of the topics
             auto & topic_callback_vec = waiting_map_.at(node);
             for (auto it = topic_callback_vec.begin(); it != topic_callback_vec.end(); ++it) {
               const std::string & topic = it->first;
               const auto callback = it->second;
-              auto opt_qos = this->get_topic_qos(topic, node);
+              std::optional<QosMatchInfo> opt_qos;
+              try {
+                opt_qos = this->get_topic_qos(topic, node);
+              } catch (const rclcpp::exceptions::RCLError & ex) {
+                // If the context was shutdown, then exit cleanly
+                // This can happen if we get a SIGINT
+                const auto context = node->get_node_options().context();
+                if (!context->is_valid()) {
+                  return;
+                }
+                throw ex;
+              }
```

Comment:
argh... this doesn't look pretty nice

Reply:
Yeah, I know.. I think the only other way around it would be to disable the default signint handler and implement our own, but that wouldn't look pretty either.